File: src/AdoubleModule.jl
module AdoubleModule
    using ADOLC_jll
    using CxxWrap
    
    # one need to specify the location of adolc_wrap.{so, dylib}
    #total_build_dir = joinpath(ADOLC_jll.artifact_dir, "lib64")
    #@wrapmodule(() -> joinpath(total_build_dir,"libadolc_wrap"), :Adouble_module)
    @wrapmodule(() -> libadolc_wrap, :Adouble_module)
   
    function __init__()
        @initcxx
        println("init adouble")
    end

# convient inits for vector of independant and dependant 
function Base.:<<(a::Vector{AdoubleModule.AdoubleCxxAllocated}, x::AbstractVector)
  for i in eachindex(x)
      a[i] << x[i]
  end
end


function Base.:>>(a::Vector{AdoubleModule.AdoubleCxxAllocated}, x::AbstractVector)
  for i in eachindex(x)
      a[i] >> x[i]
  end
end

function Base.:>>(a::AdoubleModule.AdoubleCxxAllocated, x::Vector{Float64})
  if length(x) != 1
    throw("DimensionMismatch: Length of x ($x) should be 1!")
  end
  return [a] >> x
end

function Base.:>>(a::Vector{AdoubleModule.AdoubleCxxAllocated}, x::Float64)
  if length(a) != 1
    throw("DimensionMismatch: Length of a ($a) should be 1!")
  end
  return a >> [x]
end


function Base.:*(a::AdoubleModule.AdoubleCxxAllocated, x::AbstractVector)
  return map((x_i)->a*x_i, x)
end


  function gradient(tape_num::Int64, x::Vector{Float64})
    num_ind = length(x)
    g = Vector{Float64}(undef, num_ind)
    gradient(tape_num, num_ind, x, g)
    return g
  end

  function gradient(tape_num::Int64, num_ind::Int64, x::Vector{Float64})
    g = Vector{Float64}(undef, num_ind)
    gradient(tape_num, num_ind, x, g)
    return g
  end


  export AdoubleCxx, getValue

  # general adolc
  export trace_on, trace_off, ad_forward, ad_reverse, gradient

  # more low level function
  export zos_forward, fos_forward, hos_forward, fov_forward, hov_forward
  export              fos_reverse, hos_reverse, fov_reverse, hov_reverse

  # easy to use higher order driver
  export tensor_eval, tensor_address

  # more drivers
  export   jacobian, hessian, vec_jac, jac_vec, hess_vec, hess_mat, lagra_hess_vec, jac_solv


  # point-wise smooth utils 
  export enableMinMaxUsingAbs, get_num_switches, zos_pl_forward, fos_pl_forward, fov_pl_forward, abs_normal
  # current base operations:
  # max, abs, exp, sqrt, *, +, -, ^

end # module adouble

File: src/AdoubleModule.cpp
#include <jlcxx/jlcxx.hpp>
#include <adolc/adolc.h>
#include <iostream>

adouble add(adouble const &a, adouble const &b)
{
  return a + b;
}

adouble add_right(adouble const &a, double const v)
{
  return a + v;
}

adouble add_left(double const v, adouble const &a)
{
  return v + a;
}

adouble diff(adouble const &a, adouble const &b)
{
  return a - b;
}

adouble diff_right(adouble const &a, double const v)
{
  return a - v;
}

adouble diff_left(double const v, adouble const &a)
{
  return v - a;
}
adouble diff_unary(adouble const &a)
{
  return (-1) * a;
}

adouble mult(adouble const &a, adouble const &b)
{
  return a * b;
}

adouble mult_right(adouble const &a, double const v)
{
  return a * v;
}

adouble mult_left(double const v, adouble const &a)
{
  return v * a;
}

adouble div2(adouble const &a, adouble const &b)
{
  return a / b;
}

adouble div_right(adouble const &a, double const v)
{
  return a / v;
}

adouble div_left(double const v, adouble const &a)
{
  return v / a;
}

adouble fabs2(adouble const &a)
{
  return fabs(a);
}

void assign(adouble &x, double val)
{
  x <<= val;
}

double dassign(adouble &x, double &val)
{
  x >>= val;
  return val;
}
adouble power(adouble x, int n)
{
  adouble z = 1;

  if (n > 0) /* Recursion and branches */
  {
    int nh = n / 2;   /* that do not depend on  */
    z = power(x, nh); /* adoubles are fine !!!! */
    z *= z;
    if (2 * nh != n)
      z *= x;
    return z;
  } /* end if */
  else
  {
    if (n == 0) /* The local adouble z dies */
      return z; /* as it goes out of scope. */
    else
      return 1 / power(x, -n);
  } /* end else */
}

adouble sqrt2(adouble const &a)
{
  return sqrt(a);
}
adouble exp2(adouble const &a)
{
  return exp(a);
}
adouble fmax2(adouble const &a, adouble const &b)
{
  return fmax(a, b);
}
adouble fmin2(adouble const &a, adouble const &b)
{
  return fmin(a, b);
}

JLCXX_MODULE Adouble_module(jlcxx::Module &types)
{
  types.add_type<adouble>("AdoubleCxx", jlcxx::julia_type("AbstractFloat", "Base"))
      .constructor<double>();

  types.method("getValue", [](adouble &a)
               { return a.getValue(); });
  types.method("gradient", [](int tag, int n, double *x, double *g)
               { return gradient(tag, n, x, g); });
  types.method("trace_on", [](int tag)
               { return trace_on(tag); });
  types.method("trace_on", [](int tag, int keep)
               { return trace_on(tag, keep); });
  types.method("trace_off", trace_off);

  // easy to use drivers

  types.method("jacobian", jacobian);
  types.method("hessian", hessian);
  types.method("vec_jac", vec_jac);
  types.method("jac_vec", jac_vec);
  types.method("hess_vec", hess_vec);
  types.method("hess_mat", hess_mat);
  types.method("lagra_hess_vec", lagra_hess_vec);
  types.method("jac_solv", jac_solv);

  types.method("ad_forward", [](short tag, int m, int n,
                                int d,
                                int keep,
                                double **X,
                                double **Y)
               { forward(tag,
                         m,
                         n,
                         d,
                         keep,
                         X,
                         Y); });
  types.method("ad_reverse", [](short tag,
                                int m,
                                int n,
                                int d,
                                double *u,
                                double **Z)
               { reverse(tag, m, n, d, u, Z); });

  types.method("zos_forward", zos_forward);
  types.method("fos_forward", fos_forward);
  types.method("hos_forward", hos_forward);

  types.method("fov_forward", fov_forward);
  types.method("hov_forward", hov_forward);

  types.method("fos_reverse", fos_reverse);
  types.method("hos_reverse", hos_reverse);

  types.method("fov_reverse", fov_reverse);
  types.method("hov_reverse", hov_reverse);

  // easy to use higher order drivers
  types.method("tensor_address", tensor_address);
  types.method("tensor_eval", tensor_eval);

  // pointwise-smooth functions
  types.method("enableMinMaxUsingAbs", enableMinMaxUsingAbs);
  types.method("get_num_switches", get_num_switches);
  types.method("zos_pl_forward", zos_pl_forward);
  types.method("fos_pl_forward", fos_pl_forward);
  types.method("fov_pl_forward", fov_pl_forward);
  types.method("abs_normal", abs_normal);

  //--------------------

  // basic arithmetic operations
  types.set_override_module(jl_base_module);

  types.method("+", add);
  types.method("+", add_right);
  types.method("+", add_left);

  types.method("-", diff);
  types.method("-", diff_right);
  types.method("-", diff_left);
  types.method("-", diff_unary);

  types.method("*", mult);
  types.method("*", mult_left);
  types.method("*", mult_right);

  types.method("/", div2);
  types.method("/", div_left);
  types.method("/", div_right);

  types.method("<<", assign);
  types.method(">>", dassign);

  types.method("^", [](adouble x, int n)
               { return power(x, n); });

  types.method("max", [](adouble const &a, adouble const &b)
               { return fmax2(a, b); });
  types.method("min", [](adouble const &a, adouble const &b)
               { return fmin2(a, b); });
  types.method("abs", [](adouble const &a)
               { return fabs2(a); });
  types.method("sqrt", [](adouble const &a)
               { return sqrt2(a); });
  types.method("exp", [](adouble const &a)
               { return exp2(a); });
  types.unset_override_module();
}
